File: sol/2192.cpp
#include <unordered_set>
#include <deque>

struct Solution {
    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        vector<vector<int>> edgeList(n);
        vector<int> indeg(n, 0);
        for (auto &e : edges){
            edgeList[e[0]].push_back(e[1]);
            ++indeg[e[1]];
        }
        vector<unordered_set<int>> resSet(n);
        deque<int> lvl;
        for (int i = 0; i < n; ++i){
            if (indeg[i] == 0){
                lvl.push_back(i);
            }
        }
        while (!lvl.empty()){
            int x = lvl.front();
            lvl.pop_front();
            indeg[x] = -1;
            for (int nb : edgeList[x]){
                resSet[nb].insert(x);
                for (int y : resSet[x]){
                    resSet[nb].insert(y);
                }
                if (--indeg[nb] == 0){
                    lvl.push_back(nb);
                }
            }
        }
        vector<vector<int>> res(n);
        for (int i = 0; i < n; ++i){
            res[i].insert(res[i].begin(), resSet[i].begin(), resSet[i].end());
            sort(res[i].begin(), res[i].end());
        }
        return res;
    }
};

File: sol/2192/2192.py
from collections import defaultdict

class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        edgeList = defaultdict(list)
        indeg = [0] * n
        for a, b in edges:
            edgeList[a].append(b)
            indeg[b] += 1
        resSet = [set() for _ in range(n)]
        lvl = [i for i in range(n) if indeg[i] == 0]
        while len(lvl) != 0:
            newLvl = []
            for x in lvl:
                indeg[x] = -1
                for nb in edgeList[x]:
                    indeg[nb] -= 1
                    resSet[nb].update(resSet[x])
                    resSet[nb].add(x)
                    if indeg[nb] == 0:
                        newLvl.append(nb)
            lvl = newLvl
        return [sorted(s) for s in resSet]
